File: src/forge/tools/executor.py
import subprocess
from pathlib import Path

_BLOCKED_PATTERNS: list[str] = [
    "rm -rf /",
    "rm -rf ~",
    ":(){ :|:& };:",
    "dd if=/dev/zero",
    "mkfs",
    "> /dev/sda",
    "chmod 777 /",
    "chown -R",
    "sudo rm",
    "sudo dd",
]


def _is_blocked(command: str) -> bool:
    lowered = command.lower()
    return any(pattern in lowered for pattern in _BLOCKED_PATTERNS)


def _bash_exec(args: dict[str, object], workspace: Path) -> str:
    command = str(args.get("command", ""))
    timeout = int(args.get("timeout", 60))

    if not command.strip():
        return "ERROR: Empty command"

    if _is_blocked(command):
        return f"ERROR: Command blocked for safety: {command!r}"

    try:
        result = subprocess.run(
            command,
            shell=True,
            cwd=workspace,
            capture_output=True,
            text=True,
            timeout=timeout,
        )
        output = result.stdout
        if result.stderr:
            output += f"\n[stderr]\n{result.stderr}"
        output += f"\n[exit {result.returncode}]"
        if len(output) > 8000:
            output = output[:4000] + "\n... [truncated] ...\n" + output[-4000:]
        return output.strip()
    except subprocess.TimeoutExpired:
        return f"ERROR: Command timed out after {timeout}s"
    except Exception as exc:
        return f"ERROR: {exc}"


def _read_file(args: dict[str, object], workspace: Path) -> str:
    rel_path = str(args.get("path", ""))
    if not rel_path:
        return "ERROR: No path provided"

    target = (workspace / rel_path).resolve()

    try:
        target.relative_to(workspace.resolve())
    except ValueError:
        return f"ERROR: Path escapes workspace: {rel_path}"

    if not target.exists():
        return f"ERROR: File not found: {rel_path}"

    if not target.is_file():
        return f"ERROR: Not a file: {rel_path}"

    try:
        content = target.read_text(encoding="utf-8", errors="replace")
        if len(content) > 16000:
            content = content[:8000] + "\n... [truncated] ...\n" + content[-8000:]
        return content
    except Exception as exc:
        return f"ERROR reading {rel_path}: {exc}"


def _write_file(args: dict[str, object], workspace: Path) -> str:
    rel_path = str(args.get("path", ""))
    content = str(args.get("content", ""))

    if not rel_path:
        return "ERROR: No path provided"

    target = (workspace / rel_path).resolve()

    try:
        target.relative_to(workspace.resolve())
    except ValueError:
        return f"ERROR: Path escapes workspace: {rel_path}"

    try:
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(content, encoding="utf-8")
        return f"OK: Wrote {len(content)} chars to {rel_path}"
    except Exception as exc:
        return f"ERROR writing {rel_path}: {exc}"


def _list_dir(args: dict[str, object], workspace: Path) -> str:
    rel_path = str(args.get("path", "."))
    target = (workspace / rel_path).resolve()

    try:
        target.relative_to(workspace.resolve())
    except ValueError:
        return f"ERROR: Path escapes workspace: {rel_path}"

    if not target.exists():
        return f"ERROR: Path not found: {rel_path}"

    if not target.is_dir():
        return f"ERROR: Not a directory: {rel_path}"

    lines = []
    for item in sorted(target.iterdir()):
        prefix = "d" if item.is_dir() else "f"
        lines.append(f"[{prefix}] {item.name}")

    return "\n".join(lines) if lines else "(empty directory)"


def execute_tool(name: str, args: dict[str, object], workspace: Path) -> str:
    if name == "bash_exec":
        return _bash_exec(args, workspace)
    elif name == "read_file":
        return _read_file(args, workspace)
    elif name == "write_file":
        return _write_file(args, workspace)
    elif name == "list_dir":
        return _list_dir(args, workspace)
    else:
        return f"ERROR: Unknown tool '{name}'"

File: src-ts/tools/executor.ts
import { execSync } from "child_process";
import * as fs from "fs";
import * as path from "path";

const BLOCKED_PATTERNS = [
  "rm -rf /", "rm -rf ~", ":(){ :|:& };:", "dd if=/dev/zero",
  "mkfs", "> /dev/sda", "chmod 777 /", "chown -R", "sudo rm", "sudo dd",
];

function isBlocked(command: string): boolean {
  const lower = command.toLowerCase();
  return BLOCKED_PATTERNS.some(p => lower.includes(p));
}

function bashExec(args: Record<string, unknown>, workspace: string): string {
  const command = String(args["command"] ?? "");
  const timeout = Number(args["timeout"] ?? 60) * 1000;
  if (!command.trim()) return "ERROR: Empty command";
  if (isBlocked(command)) return `ERROR: Command blocked for safety: ${command}`;
  try {
    const stdout = execSync(command, { cwd: workspace, timeout, encoding: "utf8", stdio: "pipe" });
    const out = stdout.length > 8000
      ? stdout.slice(0, 4000) + "\n... [truncated] ...\n" + stdout.slice(-4000)
      : stdout;
    return out + "\n[exit 0]";
  } catch (e: any) {
    const out = (e.stdout ?? "") + (e.stderr ? `\n[stderr]\n${e.stderr}` : "");
    return out + `\n[exit ${e.status ?? 1}]`;
  }
}

function resolveInWorkspace(relPath: string, workspace: string): string | null {
  const resolved = path.resolve(workspace, relPath);
  if (!resolved.startsWith(path.resolve(workspace) + path.sep) && resolved !== path.resolve(workspace)) return null;
  return resolved;
}

function readFile(args: Record<string, unknown>, workspace: string): string {
  const relPath = String(args["path"] ?? "");
  if (!relPath) return "ERROR: No path provided";
  const target = resolveInWorkspace(relPath, workspace);
  if (!target) return `ERROR: Path escapes workspace: ${relPath}`;
  if (!fs.existsSync(target)) return `ERROR: File not found: ${relPath}`;
  if (!fs.statSync(target).isFile()) return `ERROR: Not a file: ${relPath}`;
  try {
    let content = fs.readFileSync(target, "utf8");
    if (content.length > 16000) content = content.slice(0, 8000) + "\n... [truncated] ...\n" + content.slice(-8000);
    return content;
  } catch (e: any) {
    return `ERROR reading ${relPath}: ${e.message}`;
  }
}

function writeFile(args: Record<string, unknown>, workspace: string): string {
  const relPath = String(args["path"] ?? "");
  const content = String(args["content"] ?? "");
  if (!relPath) return "ERROR: No path provided";
  const target = resolveInWorkspace(relPath, workspace);
  if (!target) return `ERROR: Path escapes workspace: ${relPath}`;
  try {
    fs.mkdirSync(path.dirname(target), { recursive: true });
    fs.writeFileSync(target, content, "utf8");
    return `OK: Wrote ${content.length} chars to ${relPath}`;
  } catch (e: any) {
    return `ERROR writing ${relPath}: ${e.message}`;
  }
}

function listDir(args: Record<string, unknown>, workspace: string): string {
  const relPath = String(args["path"] ?? ".");
  const target = resolveInWorkspace(relPath, workspace);
  if (!target) return `ERROR: Path escapes workspace: ${relPath}`;
  if (!fs.existsSync(target)) return `ERROR: Path not found: ${relPath}`;
  if (!fs.statSync(target).isDirectory()) return `ERROR: Not a directory: ${relPath}`;
  const items = fs.readdirSync(target).sort().map(name => {
    const isDir = fs.statSync(path.join(target, name)).isDirectory();
    return `[${isDir ? "d" : "f"}] ${name}`;
  });
  return items.length ? items.join("\n") : "(empty directory)";
}

export function executeTool(name: string, args: Record<string, unknown>, workspace: string): string {
  if (name === "bash_exec") return bashExec(args, workspace);
  if (name === "read_file") return readFile(args, workspace);
  if (name === "write_file") return writeFile(args, workspace);
  if (name === "list_dir") return listDir(args, workspace);
  return `ERROR: Unknown tool '${name}'`;
}
